File: extensions/pyre/chroma/ansi.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// bind the {ansi} serializers
void
pyre::py::chroma::ansi(py::module & m)
{
    // create the {ansi} submodule
    auto ansi = m.def_submodule(
        // the name of the module
        "ansi",
        // its docstring
        "serialize a color as an ANSI terminal escape sequence");

    // the reset sequence
    ansi.def(
        // the name
        "reset",
        // the serializer
        &pyre::chroma::ansi::reset,
        // the docstring
        "the sequence that returns the terminal to its default attributes");
    // a 24-bit truecolor sequence from a color
    ansi.def(
        // the name
        "rgb",
        // the serializer
        &pyre::chroma::ansi::rgb,
        // the signature
        "color"_a, "foreground"_a = true,
        // the docstring
        "a 24-bit truecolor escape sequence for {color}");
    // a 256-color-cube sequence from a color
    ansi.def(
        // the name
        "rgb256",
        // the serializer
        &pyre::chroma::ansi::rgb256,
        // the signature
        "color"_a, "foreground"_a = true,
        // the docstring
        "a 256-color-cube escape sequence for {color}");
    // a grayscale-ramp sequence from a level
    ansi.def(
        // the name
        "gray",
        // the serializer
        &pyre::chroma::ansi::gray,
        // the signature
        "level"_a, "foreground"_a = true,
        // the docstring
        "a grayscale-ramp escape sequence for {level}");
    // a 16-color sequence from a terminal-palette code
    ansi.def(
        // the name
        "csi3",
        // the serializer
        &pyre::chroma::ansi::csi3,
        // the signature
        "code"_a, "bright"_a = false,
        // the docstring
        "a 16-color escape sequence for the terminal-palette {code}");

    // all done
    return;
}


// end of file

File: packages/pyre/terminals/ANSI.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


# support
import pyre

# my base
from .Interactive import Interactive


# declaration
class ANSI(Interactive, family="pyre.terminals.ansi"):
    """
    A live terminal that renders color using ANSI control sequences
    """

    # interface
    @pyre.export
    def render(self, color):
        """
        Render {color} (a {chroma} color) as a 24-bit truecolor ANSI escape sequence
        """
        # reach the color bindings
        chroma = self._chroma()
        # without them, or without a color to render
        if chroma is None or color is None:
            # there is nothing to emit
            return ""
        # let chroma serialize the color as a truecolor escape
        return chroma.ansi.rgb(color)

    @pyre.export
    def reset(self):
        """
        The escape sequence that restores the terminal to its default attributes
        """
        # reach the color bindings
        chroma = self._chroma()
        # without them
        if chroma is None:
            # there is nothing to emit
            return ""
        # let chroma produce the reset sequence
        return chroma.ansi.reset()


# end of file
